src/main/java/org/apache/sysml/hops/estim/EstimatorMatrixHistogram.java

```diff
@@ -31,254 +31,256 @@
 		root.setSynopsis(MatrixHistogram.deriveOutputHistogram(h1, h2, ret));
-		
+
 		return ret;
 	}
-	
-	@Override 
+
+	@Override
 	public double estim(MatrixBlock m1, MatrixBlock m2) {
 		return estim(m1, m2, OpCode.MM);
 	}
-	
+
 	@Override
 	public double estim(MatrixBlock m1, MatrixBlock m2, OpCode op) {
 		MatrixHistogram h1 = new MatrixHistogram(m1, _useExcepts);
-		MatrixHistogram h2 = (m1 == m2) ? //self product
-			h1 : new MatrixHistogram(m2, _useExcepts);
+		MatrixHistogram h2 = (m1 == m2) ? // self product
+				h1 : new MatrixHistogram(m2, _useExcepts);
 		return estimIntern(h1, h2, op);
 	}
-	
+
 	@Override
 	public double estim(MatrixBlock m1, OpCode op) {
 		MatrixHistogram h1 = new MatrixHistogram(m1, _useExcepts);
 		return estimIntern(h1, null, op);
 	}
-	
+
 	private double estimIntern(MatrixHistogram h1, MatrixHistogram h2, OpCode op) {
-		double msize = (double)h1.getRows()*h1.getCols();
-		
+		double msize = (double) h1.getRows() * h1.getCols();
+
 		switch (op) {
-			case MM:
-				return estimInternMM(h1, h2);
-			case MULT:
-				return Math.min(
-					IntStream.range(0, h1.getRows()).mapToDouble(i -> (double)h1.rNnz[i]/msize * (double)h2.rNnz[i]/msize).sum(),
-					IntStream.range(0, h1.getCols()).mapToDouble(i -> (double)h1.cNnz[i]/msize * (double)h2.cNnz[i]/msize).sum());
-			case PLUS:
-				return Math.min(
-					IntStream.range(0, h1.getRows()).mapToDouble(i -> (double)h1.rNnz[i]/msize 
-						+ (double)h2.rNnz[i]/msize - (double)h1.rNnz[i]/msize * (double)h2.rNnz[i]/msize).sum(),
-					IntStream.range(0, h1.getCols()).mapToDouble(i -> (double)h1.cNnz[i]/msize 
-						+ (double)h2.cNnz[i]/msize - (double)h1.cNnz[i]/msize * (double)h2.cNnz[i]/msize).sum());
-			case EQZERO:
-				return OptimizerUtils.getSparsity(h1.getRows(), h1.getCols(),
-					(long)h1.getRows() * h1.getCols() - h1.getNonZeros());
-			case DIAG:
-				return (h1.getCols()==1) ?
-					OptimizerUtils.getSparsity(h1.getRows(), h1.getRows(), h1.getNonZeros()) :
-					OptimizerUtils.getSparsity(h1.getRows(), 1, Math.min(h1.getRows(), h1.getNonZeros()));
-			//binary operations that preserve sparsity exactly
-			case CBIND:
-				return OptimizerUtils.getSparsity(h1.getRows(),
-					h1.getCols()+h2.getCols(), h1.getNonZeros() + h2.getNonZeros());
-			case RBIND:
-				return OptimizerUtils.getSparsity(h1.getRows()+h2.getRows(),
-					h1.getCols(), h1.getNonZeros() + h2.getNonZeros());
-			//unary operation that preserve sparsity exactly
-			case NEQZERO:
-			case TRANS:
-			case RESHAPE:
-				return OptimizerUtils.getSparsity(h1.getRows(), h1.getCols(), h1.getNonZeros());
-			default:
-				throw new NotImplementedException();
+		case MM:
+			return estimInternMM(h1, h2);
+		case MULT:
+			double[] out_r = null;
+			for (int i = 0; i < h1.getRows(); i++) {
+				out_r[i] = (h1.rNnz[i] * h2.rNnz[i]
+						* IntStream.range(0, h1.getCols()).mapToDouble(k -> h1.cNnz[k] * h2.cNnz[k]).sum())
+						/ (Arrays.stream(h1.rNnz).sum() * Arrays.stream(h1.rNnz).sum());
```

Comment:
These repeated scans of the count vectors for each row in the lhs, would cause a time complexity in the number of cells of the output. Luckily, these are constants with regard to different rows and hence can be computed once.